pkg/cfn/manager/waiters.go

```diff
@@ -0,0 +1,182 @@
+	"strings"
+	"time"
+
+	"github.com/pkg/errors"
+
+	"github.com/aws/aws-sdk-go/aws/request"
+	cfn "github.com/aws/aws-sdk-go/service/cloudformation"
+
+	"github.com/kubicorn/kubicorn/pkg/logger"
+
+	"github.com/weaveworks/eksctl/pkg/cfn/builder"
+)
+
+// cloudformation.WaitUntilStackCreateComplete doesn't detect in-progress status early enough,
+// so this is custom version that is more suitable for our use, as there is no way to add any
+// custom acceptors
+
+func makeStatusAcceptor(status string) request.WaiterAcceptor {
+	return request.WaiterAcceptor{
+		Matcher:  request.PathAllWaiterMatch,
+		Argument: "Stacks[].StackStatus",
+		Expected: status,
+		State:    request.FailureWaiterState,
+	}
+}
+
+func makeAcceptors(successStatus string, failureStates []string, extraAcceptors ...request.WaiterAcceptor) []request.WaiterAcceptor {
+	acceptors := []request.WaiterAcceptor{makeStatusAcceptor(successStatus)}
+	acceptors[0].State = request.SuccessWaiterState
+
+	for _, s := range failureStates {
+		acceptors = append(acceptors, makeStatusAcceptor(s))
+	}
+
+	acceptors = append(acceptors, extraAcceptors...)
+
+	return acceptors
+}
+
+// makeWaiterDelay returns random 15s±5000ms delay
+func makeWaiterDelay() request.WaiterDelay {
+	const (
+		base        = 15 * time.Second
+		offsetSteps = 200
+		offsetMax   = 5000
+		stepMult    = offsetMax / offsetSteps
+	)
+
+	offsets := rand.Perm(offsetSteps)
+
+	return func(attempt int) time.Duration {
+		s := rand.Intn(offsetSteps)
+		d := stepMult * offsets[s]
+
+		offset := time.Duration(d) * time.Millisecond
+
+		return base + offset
+	}
+}
+
+func (c *StackCollection) waitWithAcceptors(name string, acceptors []request.WaiterAcceptor) error {
+	desiredStatus := fmt.Sprintf("%v", acceptors[0].Expected)
+	msg := fmt.Sprintf("waiting for CloudFormation stack %q to reach %q status", name, desiredStatus)
+
+	ctx, cancel := context.WithTimeout(context.Background(), c.spec.WaitTimeout)
+	defer cancel()
+
+	w := request.Waiter{
+		Name:        strings.Join([]string{"wait", name, desiredStatus}, "_"),
+		MaxAttempts: 1024, // we use context deadline instead
+		Delay:       makeWaiterDelay(),
+		Acceptors:   acceptors,
+		NewRequest: func(_ []request.Option) (*request.Request, error) {
+			input := &cfn.DescribeStacksInput{
+				StackName: &name,
+			}
+			logger.Debug(msg)
+			req, _ := c.cfn.DescribeStacksRequest(input)
+			req.SetContext(ctx)
+			return req, nil
+		},
+	}
+
+	logger.Debug("start %s", msg)
+
+	if waitErr := w.WaitWithContext(ctx); waitErr != nil {
+		s, err := c.describeStack(name)
+		if err != nil {
+			logger.Critical("unexpected status %q while %s", *s.StackStatus, msg)
+		} else {
+			logger.Debug("describeErr=%v", err)
+		}
+		return errors.Wrap(waitErr, msg)
+	}
+
+	logger.Debug("done %s", msg)
+
+	return nil
+}
+
+func (c *StackCollection) doWaitUntilStackIsCreated(name string) error {
+	return c.waitWithAcceptors(name,
+		makeAcceptors(
+			cfn.StackStatusCreateComplete,
+			[]string{
```

Comment:
It unfortunate we have to list every stack status we want to error on, but I'd rather keep this code simple and not try to make it do anything too clever.